type-metadata/src/lib.rs

```diff
@@ -1,7 +1,319 @@
-#[cfg(test)]
-mod tests {
-	#[test]
-	fn it_works() {
-		assert_eq!(2 + 2, 4);
+// Copyright 2019 Centrality Investments Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+mod registry;
+pub use registry::Registry;
+
+mod tests;
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct TypeIdent {
+	pub namespace: Vec<String>,
+	pub ident: IdentKind,
+	pub args: Vec<TypeIdent>,
+}
+
+impl TypeIdent {
+	pub fn new(ident: IdentKind) -> TypeIdent {
+		TypeIdent {
+			namespace: vec![],
+			ident,
+			args: vec![],
+		}
+	}
+	pub fn namespace(mut self, namespace: Vec<String>) -> Self {
+		self.namespace = namespace;
+		self
+	}
+	pub fn args(mut self, args: Vec<TypeIdent>) -> Self {
+		self.args = args;
+		self
+	}
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub enum IdentKind {
+	Custom(String),
+
+	// primitives, including build-in and common core/std primitives
+	Bool,
+	Str,
+	Unit,
+	U8,
+	U16,
+	U32,
+	Usize,
+	U64,
+	U128,
+	I8,
+	I16,
+	I32,
+	Isize,
+	I64,
+	I128,
+	Array(u16),
+	Vector,
+	Tuple,
+	Option,
+	Result,
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum TypeDef {
+	Primitive,
+	Struct(StructDef),
+	EnumDef(EnumDef),
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum FieldName {
+	Unnamed(u16),
+	Named(String),
+}
+#[derive(PartialEq, Eq, Debug)]
+pub struct Field {
+	pub name: FieldName,
+	pub ident: TypeIdent,
+}
+pub type StructDef = Vec<Field>;
+
+#[derive(PartialEq, Eq, Debug)]
+pub struct EnumVariant {
+	pub name: String,
+	pub index: u16,
+	pub fields: Vec<Field>,
+}
+pub type EnumDef = Vec<EnumVariant>;
+
+pub trait Metadata {
+	fn type_ident() -> TypeIdent;
+
+	/// `type_def` would do registration if needed
+	fn type_def(registry: &mut Registry) -> TypeDef;
+
+	fn register(registry: &mut Registry) {
+		registry.register(Self::type_ident(), Self::type_def);
+	}
+}
+
+macro_rules! impl_metadata_for_primitives {
+	( $( $t:ty => $ident_kind:expr, )* ) => { $(
+		impl Metadata for $t {
+			fn type_ident() -> TypeIdent {
+				TypeIdent::new($ident_kind)
+			}
+
+			fn type_def(_registry: &mut Registry) -> TypeDef {
+				TypeDef::Primitive
+			}
+		}
+	)* }
+}
+
+impl_metadata_for_primitives!(
+	bool => IdentKind::Bool,
+	u8 => IdentKind::U8,
+	u16 => IdentKind::U16,
+	u32 => IdentKind::U32,
+	usize => IdentKind::Usize,
```

Comment:
Some types are either missing or there are types that shouldn't be there.
For example I am having a hard time figuring out why we should serialize `usize` and `isize` types.
They cannot be safely shared between different hosts anyway.

Also is this a general framework or specialized for substrate?
Depending on this we also might want types like `char`.

Reply:
I believe our current scope is a specialized one for ink! and substrate, though we could adapt it to a more general framework in the future. Btw I've removed `usize` and `isize`.